stempy/reader.h

```diff
@@ -444,6 +448,165 @@ std::future<void> SectorStreamThreadedReader::readAll(Functor& func)
 
   return complete;
 }
+
+// struct to hold the location of a sector: sector, stream and offset
+struct SectorLocation
+{
+  uint64_t sector = -1;
+  SectorStreamReader::SectorStream* sectorStream = nullptr;
+  std::streampos offset;
+};
+
+// struct FrameMap {
+//   SectorLocation locations[4];
+// };
+
+// This type holds the locations of the sectors for each frame at a give scan
+// position. Note: That there can be multiple frames at a give position, we key
+// the map of the frame number.
+using ScanMap = std::vector<std::map<uint32_t, std::array<SectorLocation, 4>>>;
+
+// SectorStreamMultiPassThreadedReader uses a two pass approach to frame
+// reconstruction. First is reads the header from all sectors and uses them
+// to build up "frame maps" detailing the streams and offsets for all the
+// sectors in a frame. The second pass is used reconstruct the frames using this
+// map. This avoids the need to cache partial frames so used less memory.
+// Depending on the seek performance of the disks it also performs well.
+class SectorStreamMultiPassThreadedReader : public SectorStreamThreadedReader
+{
+public:
+  SectorStreamMultiPassThreadedReader(const std::string& path);
+  SectorStreamMultiPassThreadedReader(const std::vector<std::string>& files);
+  SectorStreamMultiPassThreadedReader(const std::string& path, int threads = 0);
```

Comment:
I think the default argument `threads = 0` would never get used, since the single argument `SectorStreamMultiPassThreadedReader(const std::string& path);` would get preference in overload resolution.

Would it make sense to just have a single `SectorStreamMultiPassThreadedReader(const std::string& path, int threads = -1);` constructor? Also, similar kind of thing for the vector version.

Reply:
Yes, that makes sense, I have pushed a fix.